### DISCUSSION TITLE
pTx extensions for parallel transmit - existing implementations and path into the future

### DISCUSSION THREAD

**Discussion by dabosch** [pulseq/pulseq#69] [Ideas]
Dear all,

since we've already discussed it with a few people I want to use this topic to summarize the status of parallel transmit in pulseq, and to then draft a roadmap of what could be done.

## Current status of pTx pulseq

The first version of pTx-pulseq I am aware of was a version I hacked together back in ~2020, but after initial tests I never used it due to too many other projects. This version would simply use a external.seq file for each channel, named external_0.seq, external_1.seq, ... 
This was mostly a demo to show that pulseq and pTx can be used together, and as far as I know, this was only slightly used by Moritz Zaiss from Erlangen and Sebastian Schmitter from Berlin.

In 2023, Martin Freudensprung from Erlangen published results from his pulseq extension for RF shimming, enabling composite pulses [\[1\]][1]. This works well and I have used it myself [\[2\]][2]. It does, however, have a few drawbacks, as it is incompatible with previous .seq files. Also, the Siemens systems have a minimum delay of 100µs between consecutive pulses, which can be a problem for composite pulses.

At ISMRM 2024, Thomas Roos @Roosted7 published his work of bringing pulseq to Philips, including pTx support [\[3\]][3]. This implementation uses what I would call an elegant hack. Firstly, all RF pulseq shapes of the individual channels are concatenated. The time shape is then replicated once for each channel, resulting in a time shape vector that resets to 0 multiple times. A very convenient aspect of this method is that it is compatible with the existing matlab and python toolboxes, including the plotting tools (yes, they plot all channels by default!). When using an unmodified pulsesq interpreter, the waveform of the first channel will be played out.

I then started working on the Siemens pulseq interpreter and implemented support for pTx pulses, following Thomas' method. I am now at a point where it works reliably and I have been able to use it on a few projects, both on Syngo VE12U and on Syngo XA70. The interpreter does recognize the difference between RF pulses which just have a RF shim and treats them differently than pulses with different waveforms for the channels for increased performance. In this version, I have not yet included support for VB17 systems, but if there is demand I can do that.

## To Do

I think one of the first decisions should be: is the pTx pulse part of the sequence, and should therefore live in the .seq file? This is convenient since only one file needs to be transferred, and makes it possible to reproduce your experiment with this exact .seq file. On the other hand, the .seq files are then dependent on the RF coil, and if no universal pulses are used they also are subject specific.
I advocate for keeping the pulses in the .seq file, but I think it should be discussed if this is consensus in the community or if the pulses better be kept in a separate file.
Maybe we could even include multiple RF pulses for different RF coils into the same file. The interpreter would then have to chose which pulse to use depending on the coil connected to the system. I am not sure how this could be done in the .seq files without causing a huge mess there though.

Another aspect is the format of the pulses: I like the approach by Thomas, since it is compatible with the existing tools. It does, however, follow a certain logic that might not be intuitive to everybody. One drawback of this approach is that RF pulses with RF shims are replicated for each channel, even if the waveform is the same except for the phase and magnitude. I do not consider this a huge drawback though.

Lastly, we should discuss how to share the pTx compatible version of the pulseq Siemens interpreter. While my version is stable, I would not give it to anybody without warning. For all I know it behaves exactly the same as the original code for non-pTx .seq files, but I cannot make definite promises. pTx functionality is fairly well tested, but certainly not as good as the original code.
I am planning to publish some work with it and am happy to share it as openly as I can soon, but until then I'd prefer to only share it upon request with some strings attached (such as: if you publish with it, please acknowledge the massive amount of work I had to sink into this project).

## Final Words
- As mentioned above, we can discuss sharing the current status for those of you who are impatient.
- I am looking forward to your opinions and to a fruitful discussion, especially to what @m-a-x-i-m-z and @QingpingChen1201 think, since they know the pulseq project better than anybody else.

May the FID be with you!
Dario

P.S.: does anybody have a 3T system which supports free-waveform pTx? There are rumors that they exist but I haven't had the luck to work with one. I'd like to test the interpreter on one though.

### References:
[1]: https://hdl.handle.net/21.11116/0000-000D-D793-A
\[1\] ESMRMB 2023, LT53. https://hdl.handle.net/21.11116/0000-000D-D793-A

[2]: https://submissions.mirasmart.com/ISMRM2024/Itinerary/EventDetail.aspx?evt=253
\[2\] ISMRM 2024, Program Number 1032

[3]: https://submissions.mirasmart.com/ISMRM2024/Itinerary/EventDetail.aspx?evt=326
\[3\] ISMRM 2024, Program Number 1149

**Comment by mzaiss** [comment]
thanks for the nice overview, Dario!

I also think it should be in the pulseq file so the ptx sequence
idea/realization is described by the seq file alone. Do you think we could
combine the benefits of the Roos and the Freudensprung implementation?
like use one for shim and the other fir full ptx?

Dario Bosch ***@***.***> schrieb am Mi., 27. Nov. 2024, 08:11:

**Comment by dabosch** [comment]
I think we could come up with a way to encode RF shims differently from the waveforms, which would significantly reduce file size, especially when the same pulse shape is used with different shims.
I have a couple of ideas how this could be done with my current implementation. We should definitely keep this point in mind for future development.

**Comment by mzaiss** [comment]
for this we should also know how different vendors treat B1shims vs.
dynamic ptx.
especially if it is used jointly.

Dario Bosch ***@***.***> schrieb am Do., 28. Nov. 2024, 08:09:

**Comment by m-a-x-i-m-z** [comment]
The August 2025 release (v1.5.1) contains an RF shim extension, which is an invitation for the community to try to use it (also on different platforms) to move towards some kind of consensus.

**Comment by mzaiss** [comment]
Maybe we can collect more discussion points for pTX-Pulseq here:

Needs:
    one solution
    or two?

Open Problems:
    Display of B1 shims in seq.plot()
    Static vs. dynamic solutions, static shims (per pulse) vs. full pTX
